### PR TITLE
25 update user api

### LINKED ISSUE #25: Create an update user API
Create an update user API to be used for editing a users hours and/or user information

# Acceptance Criteria
- New API `PATCH /users`
- Unit Tests

### PR DESCRIPTION
**Pull request by bchampp**
Closes #25

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (bchampp) [inline .vscode/launch.json:1]: delete

File: api/src/constants/errors.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ export class ErrorConstants {
     public static readonly VALIDATION_BODY_EMAIL = "The request body is missing the email field.";
     public static readonly VALIDATION_BODY_STUDENTNUMBER = "The request body is missing the studentNumber field.";
     public static readonly VALIDATION_BODY_USERID = "The request body is missing the userId field.";
+    public static readonly VALIDATION_BODY_CHECKIN = "The request body is missing the check in field";
+    public static readonly VALIDATION_BODY_CHECKOUT = "The request body is missing the check out field";
```
Comment (bchampp) [inline api/src/constants/errors.ts:10]: should say `checkIn` and `checkOut` to be consistent with the other errors. This makes it easier to debug

File: api/src/handlers/hours.ts [hunk 3]
```diff
@@ -68,3 +65,69 @@ export const checkIn = async (
...
 };
 
+/**
+ * Update a student's hours based on their student number, add the transaction and update their hours.
+ *
+ * @param event The APIGatewayProxyEvent for the API.
+ * @returns The updated user object.
+ */
+
+// TODO: decrease hours?
```
Comment (bchampp) [inline api/src/handlers/hours.ts:75]: maybe post a ticket about this ?

File: api/src/handlers/hours.ts [hunk 4]
```diff
@@ -68,3 +65,69 @@ export const checkIn = async (
...
+      ":v_title": data.studentNumber,
+    },
+    ScanIndexForward: false,
+  };
+  try {
+    const userList = await DynamoUtilities.query(params, dynamoDb);
+
+    if (userList.length != 1) {
+      throw new Error(ErrorConstants.DYNAMO_NONUNIQUE_STUDENTNUMBER);
+    }
```
Comment (bchampp) [inline api/src/handlers/hours.ts:117]: I'm thinking that we could move this logic to the dynamo wrapper instead, and make the `query` more specific like `queryUser`. It's going to be reused anywhere we do a query based on student number.

File: api/src/types/requests/UpdateHoursRequest.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { HourTransaction } from './../HourTransaction';
+
+/**
+ * Hours Request Object for updating hours
+ */
+ export interface UpdateHoursRequest extends HourTransaction {
```
Comment (bchampp) [inline api/src/types/requests/UpdateHoursRequest.ts:6]: clean

File: api/src/util/hoursUtilities.ts [hunk 6]
```diff
@@ -48,4 +44,25 @@ export class HoursUtilities {
...
+    let totalHours = 0;
+    newUser.transactions.forEach((el) => {
+      totalHours += HoursUtilities.calculateHours(el.checkIn, el.checkOut);
+    });
+    newUser.hours = totalHours;
+
+    return newUser;
+  };
+
+  public static calculateHours = (checkIn: string, checkOut: string): number => {
```
Comment (bchampp) [inline api/src/util/hoursUtilities.ts:65]: 👍

File: api/src/handlers/hours.ts (hunk 4 above)
Comment (avalittle) [reply api/src/handlers/hours.ts:117]: did you do this?

Comment (bchampp) [reply api/src/handlers/hours.ts:117]: nah

**Pull request merged**